### PR TITLE
fix: browser back button now updates tab on coderofthemonth pages (#9…

### LINKED ISSUE #9677: [BUG] Browser back button does not update displayed tab on Coder of the Month pages
## Checklist

- [x] I have searched existing issues and confirmed this is not a duplicate (Required)

## Current Behavior

On both the **"Overall Coder of the Month"** (`/coderofthemonth/`) and **"Overall Female Coder of the Month"** (`/coderofthemonth/?category=female`) pages, when switching between tabs (e.g., "Coder of the Month" → "Top 100…" → "List of candidates…"), each tab click correctly updates the URL hash and displays the corresponding content. However, when pressing the browser **Back** button, the URL hash changes back to the previous tab but **the displayed content remains stuck on the last viewed tab**. The UI does not update to reflect the hash change.

## Expected Behavior

Pressing the browser Back button should navigate to the previously viewed tab, updating both the URL hash **and** the displayed tab content to match the same behavior that was recently fixed for the profile page in #9396 (which fixed #9392).

## If necessary, add the following information (otherwise delete this section):

https://github.com/user-attachments/assets/b03cc640-38c5-4be3-8933-70657ece6915

1. Navigate to `/coderofthemonth/`
2. Click on the **"Top 100…"** tab  URL changes to `/coderofthemonth/#codersOfPreviousMonth`, content updates correctly
3. Click on the **"List of candidates…"** tab URL changes to `/coderofthemonth/#candidatesToCoderOfTheMonth`, content updates correctly
4. Press the browser **Back** button URL changes to `/coderofthemonth/#codersOfPreviousMonth`, but content **stays on "List of candidates…"**
5. Press **Back** again URL changes to `/coderofthemonth/#codersOfTheMonth`, but content **still stays on "List of candidates…"**
6. Press **Back** once more navigates away from the page entirely

The same steps reproduce on `/coderofthemonth/?category=female`.

_Relevant details about the environment in which you encountered this bug._

* Browser name (e.g., Chrome 39, node.js 5.4): Chrome (reproducible on all browsers)
* Operating System and version (desktop or mobile): Windows (desktop)
* Link to the site where this issue occurs: `/coderofthemonth/` and `/coderofthemonth/?category=female`

---

**Affected files:**
- `frontend/www/js/omegaup/coderofthemonth/index.ts`
- `frontend/www/js/omegaup/components/coderofthemonth/List.vue`

### PR DESCRIPTION
**Pull request by manishpatel00**
# Description

Fix browser back/forward button not updating the displayed tab content on 
`/coderofthemonth/` and `/coderofthemonth/?category=female` pages.

Fixes: https://github.com/omegaup/omegaup/issues/9677

## Problem

When users press the browser Back button after viewing different tabs, the URL 
hash updates correctly but the displayed tab content remains stuck on the 
previously viewed tab.

Example:
- Click "Candidates" tab → URL: #candidatesToCoderOfTheMonth, Content: Candidates
- Click "Top 100" tab → URL: #codersOfPreviousMonth, Content: Top 100
- Press Back → URL: #candidatesToCoderOfTheMonth, Content: **still shows Top 100**

## Root Cause

No `hashchange` event listener existed in `index.ts`. When the browser Back button 
is pressed, `window.location.hash` updates automatically but the Vue component's 
`currentSelectedTab` is never notified of the change.

## Solution

Applied the proven reactive synchronization pattern from `user/profile.ts` 
(PR #9396, which fixed identical issue #9392):

1. **index.ts** - Moved `selectedTab` into reactive `data()` so Vue can propagate 
   changes to child component
2. **index.ts** - Added `hashchange` event listener that updates `selectedTab` 
   when browser Back/Forward is pressed
3. **List.vue** - Added `@Watch('selectedTab')` decorator to sync `currentSelectedTab` 
   when the prop changes from parent

### Before Fix
Browser back button changes URL but content stays on previous tab (not working):

https://github.com/user-attachments/assets/0d34fa82-2aca-4087-abce-966daa9c957b

### After Fix
Browser back button updates both URL and tab content correctly:

https://github.com/user-attachments/assets/86c4c9d1-b485-452e-9744-112f5a1d0f82

## Changes

- `frontend/www/js/omegaup/coderofthemonth/index.ts` (34 insertions, 9 deletions)
- `frontend/www/js/omegaup/components/coderofthemonth/List.vue` (5 insertions)

Both `/coderofthemonth/` and `/coderofthemonth/?category=female` are fixed with 
this single change since both use the same component.

## Verification

- ESLint: No errors
- Build: yarn build passed (47s + 5s)
- Tests: 729/729 tests passed
- Back button now correctly updates both URL and tab content

Fixes: #9677

# Comments

Reference: Identical fix pattern used in `user/profile.ts` line 562 
(`window.addEventListener('hashchange', ...)`) which resolved issue #9392 
via PR #9396.

# Checklist:

- [x] The code follows the coding guidelines of omegaUp.
- [x] The tests were executed and all of them passed.
- [x] The change is small and focused — single PR sufficient.
- [x] No breaking changes to existing functionality.
- [x] Backward compatible with all existing code.

### CODE REVIEW

File: frontend/www/js/omegaup/coderofthemonth/index.ts [hunk 1]
```diff
@@ -55,6 +58,16 @@ OmegaUp.on('ready', () => {
     },
   });
 
+  // Handle browser back/forward button navigation
+  // Same pattern used in frontend/www/js/omegaup/user/profile.ts
```
Comment (Yashagarwal9798) [inline frontend/www/js/omegaup/coderofthemonth/index.ts]: The second comment where this pattern comes from, but it'd be more useful to explain *why* the listener is needed. Consider:

```suggestion
  // Handle browser back/forward button navigation
  // The browser updates the URL hash automatically, but Vue's selectedTab
  // needs to be synced manually to reflect the correct tab in the UI.
```
Rest of the changes LGTM , final call will be taken by @pabo99 .

**Review by pabo99 — COMMENTED** [review]
In addition to @Yashagarwal9798’s suggestion, consider cleaning up the listener before destroying the Vue instance to avoid potential memory leaks

**Comment by manishpatel00** [comment]
Hii @pabo99 

**Memory leak prevention has been added** per review feedback. Fixed browser back button not updating displayed tab content on Coder of the Month pages. 

**Fixes:** #9677

---

### Changes Made

**frontend/www/js/omegaup/coderofthemonth/index.ts**
- Moved `selectedTab` into reactive `data()` for Vue component state propagation
- Added `hashchange` event listener to synchronize tab when browser navigation occurs
- **NEW: Memory leak prevention** - Event listener properly cleaned up via `$on('hook:destroyed')` hook (addresses @pabo99's feedback)

**frontend/www/js/omegaup/components/coderofthemonth/List.vue**
- Added `@Watch('selectedTab')` decorator to sync `currentSelectedTab` prop when parent updates

---

### Verification & Testing

**All 729 tests pass**

File: frontend/www/js/omegaup/coderofthemonth/index.ts (hunk 1 above)
Comment (pabo99) [reply frontend/www/js/omegaup/coderofthemonth/index.ts:69]: It doesn't appear to be addressed yet. Mentioning where the pattern comes from is less useful than explaining why synchronization is necessary

File: frontend/www/js/omegaup/components/coderofthemonth/List.vue [hunk 2]
```diff
@@ -77,6 +77,11 @@ export default class CoderOfTheMonthList extends Vue {
   T = T;
   currentSelectedTab = this.selectedTab;
 
+  @Watch('selectedTab')
```
Comment (pabo99) [inline frontend/www/js/omegaup/components/coderofthemonth/List.vue:80]: Could you add a regression test for this behavior? At minimum, the component test should verify that changing the `selectedTab` prop updates `currentSelectedTab`, the active tab, and the rendered list. Ideally, an integration test would also dispatch a `hashchange` event to cover the original Back/Forward navigation bug

File: frontend/www/js/omegaup/coderofthemonth/index.ts [hunk 3]
```diff
@@ -55,6 +65,21 @@ OmegaUp.on('ready', () => {
...
+  const onHashChange = () => {
+    const hash = window.location.hash.substring(1).split('/')[0];
+    const validTab = getSelectedValidTab(hash);
+    coderOfTheMonthList.selectedTab = validTab;
+  };
+
+  window.addEventListener('hashchange', onHashChange);
+
+  // Clean up event listener on Vue instance destruction to prevent memory leaks
+  coderOfTheMonthList.$on('hook:destroyed', () => {
```
Comment (pabo99) [inline frontend/www/js/omegaup/coderofthemonth/index.ts:79]: Other pages with the same pattern use `$once('hook:beforeDestroy', ...)`, which removes the listener before teardown and makes the one-time intent explicit. Could you use that convention here as well?

**Review by pabo99 — COMMENTED** [review]
Thanks for updating the PR. The previous memory leak concern has been addressed by keeping a stable reference to the `hashchange` handler and removing it when the `Vue` instance is destroyed.

After applying the suggested changes, could you run:
```bash
  ./stuff/lint.sh fix -- frontend/www/js/omegaup/coderofthemonth/index.ts
``` 
in order to fix the lint errors?
